Sources/SourceKitLSP/Swift/SyntaxHighlightingToken.swift

```diff
@@ -236,8 +228,35 @@ extension Array where Element == SyntaxHighlightingToken {
   /// preferring the given array's tokens if duplicate ranges are
   /// found.
   public func mergingTokens(with other: [SyntaxHighlightingToken]) -> [SyntaxHighlightingToken] {
-    let otherRanges = Set(other.map(\.sameLineRange))
-    return filter { !otherRanges.contains($0.sameLineRange) } + other
+    let otherRanges = Set(other.map(\.range))
+    return filter { !otherRanges.contains($0.range) } + other
+  }
+}
+
+extension Range where Bound == Position {
+  /// Splits a potentially multi-line range to multiple single-line ranges.
+  fileprivate func splitToSingleLineRanges(in snapshot: DocumentSnapshot) -> [Self] {
+    guard let startIndex = snapshot.index(of: lowerBound),
```

Comment:
Do you think it would be worth adding a special fast case if the range is already on the same line. E.g.
```swift
if lowerBound.line == upperBound.line {
  return [self]
}
```

That would save us the performance overhead of all the string manipulation for the common case.